src/algebra/monoid_algebra.lean

```diff
@@ -72,6 +74,75 @@ instance : has_mul (monoid_algebra k G) :=
 lemma mul_def {f g : monoid_algebra k G} :
   f * g = (f.sum $ λa₁ b₁, g.sum $ λa₂ b₂, single (a₁ * a₂) (b₁ * b₂)) :=
 rfl
+/-- The unit of the multiplication is `single 1 1`, i.e. the function
+  that is `1` at `1` and zero elsewhere. -/
+instance : has_one (monoid_algebra k G) :=
+⟨single 1 1⟩
+
+lemma one_def : (1 : monoid_algebra k G) = single 1 1 :=
+rfl
+
+instance : semiring (monoid_algebra k G) :=
+{ one       := 1,
+  mul       := (*),
+  zero      := 0,
+  add       := (+),
+  one_mul   := assume f, by simp only [mul_def, one_def, sum_single_index, zero_mul,
+    single_zero, sum_zero, zero_add, one_mul, sum_single],
+  mul_one   := assume f, by simp only [mul_def, one_def, sum_single_index, mul_zero,
+    single_zero, sum_zero, add_zero, mul_one, sum_single],
+  zero_mul  := assume f, by simp only [mul_def, sum_zero_index],
+  mul_zero  := assume f, by simp only [mul_def, sum_zero_index, sum_zero],
+  mul_assoc := assume f g h, by simp only [mul_def, sum_sum_index, sum_zero_index, sum_add_index,
+    sum_single_index, single_zero, single_add, eq_self_iff_true, forall_true_iff, forall_3_true_iff,
+    add_mul, mul_add, add_assoc, mul_assoc, zero_mul, mul_zero, sum_zero, sum_add],
+  left_distrib  := assume f g h, by simp only [mul_def, sum_add_index, mul_add, mul_zero,
+    single_zero, single_add, eq_self_iff_true, forall_true_iff, forall_3_true_iff, sum_add],
+  right_distrib := assume f g h, by simp only [mul_def, sum_add_index, add_mul, mul_zero, zero_mul,
+    single_zero, single_add, eq_self_iff_true, forall_true_iff, forall_3_true_iff, sum_zero,
+    sum_add],
+  .. finsupp.add_comm_monoid }
+
+end semiring
+
+/-! ### Derived instances -/
+section derived_instances
+
+instance [comm_semiring k] [comm_monoid G] : comm_semiring (monoid_algebra k G) :=
```

Comment:
This is not a derived instance.

Reply:
I'm using "derived" loosely here. In this case, I mean "properties which come from additional type classes on G and k ".

Reply:
Are you fine with me merging as is, or do you want me to move something here?

Reply:
For me "derived" means "something `@[derive]` should be able to do even if it can't do it now". I'd move `comm_semiring` right above the `Derived instances` section title but you may merge it either way.

Reply:
Moved